### PR TITLE
[Refactor] Deprecate Expr::operator=

### PR DESCRIPTION
**Pull request by junaire**
Signed-off-by: Jun <jun@junz.org>

Related issue = #2684

### CODE REVIEW

File: taichi/ir/expr.h [hunk 1]
```diff
@@ -89,7 +91,7 @@ class Expr {
   Expr eval() const;
 
   template <typename T, typename... Args>
-  static Expr make(Args &&... args) {
```
Comment (junaire) [inline taichi/ir/expr.h:92]: It's interesting that my editor automatically format this through clang-format...

File: taichi/ir/expr.cpp [hunk 2]
```diff
@@ -84,6 +84,10 @@ Expr &Expr::operator=(const Expr &o) {
   } else {
     set(o);  // Literally set this Expr to o
   }
+}
+
+Expr &Expr::operator=(const Expr &o) {
```
Comment (k-ye) [inline taichi/ir/expr.cpp:89]: Any chance we can remove this completely :-)

Comment (junaire) [reply taichi/ir/expr.cpp:89]: I'm afraid not? When mark this as `=delete`, we can't compile taichi.

https://github.com/taichi-dev/taichi/blob/master/taichi/ir/expression.h#L89

```
9: error: overload resolution selected deleted operator '='
      *__position = std::forward<_Arg>(__arg);
      ~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/vector.tcc:154:6: note: in instantiation of function template specialization 'std::vector<taichi::lang::Expr, std::allocator<taichi::lang::Expr> >::_M_insert_aux<taichi::lang::Expr>' requested here
            _M_insert_aux(__pos, std::move(__x_copy._M_val()));
            ^
/home/jun/dev/cpp-projects/taichi/taichi/ir/expression.h:89:11: note: in instantiation of member function 'std::vector<taichi::lang::Expr, std::allocator<taichi::lang::Expr> >::insert' requested here
    exprs.insert(exprs.begin(), a);
          ^
/home/jun/dev/cpp-projects/taichi/taichi/ir/expr.h:78:9: note: candidate function has been explicitly deleted
  Expr &operator=(const Expr &o) = delete;
```
STL isn't happy about it...

Comment (k-ye) [reply taichi/ir/expr.cpp:89]: Oh sorry, I meant somerthing like `operator=(...) = default` (I.e. The thing that annoys me is the fact that `operator=` doesn't follow its canonical meaning..)

Comment (junaire) [reply taichi/ir/expr.cpp:89]: I will try to work it out :-)

Comment (junaire) [reply taichi/ir/expr.cpp:89]: > Oh sorry, I meant somerthing like `operator=(...) = default` (I.e. It's fine to keep the `operator=`. The thing that annoys me is the fact that `operator=` doesn't follow its canonical meaning...)

Got it!

**Comment by k-ye** [comment]
Hmm, a few test failed. I guess this PR doesn't really cover all the usage of `operator=`

**Comment by junaire** [comment]
> Hmm, a few test failed. I guess this PR doesn't really cover all the usage of `operator=`

It is very interesting... Because I found all usages by following `go to reference`, maybe there are any indirect calls? I wonder code like this whether break or not:
```
ExprGroup(const Expr &a, const ExprGroup &b) {
    exprs = b.exprs;
    exprs.insert(exprs.begin(), a);
  }
```
`vector::insert` calls `Expr::operator=` internally.

**Comment by k-ye** [comment]
> > Hmm, a few test failed. I guess this PR doesn't really cover all the usage of `operator=`
> 
> It is very interesting... Because I found all usages by following `go to reference`, maybe there are any indirect calls? I wonder code like this whether break or not:
> 
> ```
> ExprGroup(const Expr &a, const ExprGroup &b) {
>     exprs = b.exprs;
>     exprs.insert(exprs.begin(), a);
>   }
> ```
> 
> `vector::insert` calls `Expr::operator=` internally.

Ah, this could very likely be the reason (https://stackoverflow.com/a/3272731/12003165 says that "as of C++11, standard containers all use copy/move constructors for non-assignment operations (like push_back)", but doesn't say how vector's `operator=` gets implemented). Pre c++11 does require the value type to be copy-assignable/constructible (https://en.cppreference.com/w/cpp/container/vector)... I guess a simple test case can help us figure out whether `T::operator=` is used. If so, could you try another way to implement `exprs = b.exprs`?

**Comment by junaire** [comment]
> > > Hmm, a few test failed. I guess this PR doesn't really cover all the usage of `operator=`
> > 
> > 
> > It is very interesting... Because I found all usages by following `go to reference`, maybe there are any indirect calls? I wonder code like this whether break or not:
> > ```
> > ExprGroup(const Expr &a, const ExprGroup &b) {
> >     exprs = b.exprs;
> >     exprs.insert(exprs.begin(), a);
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `vector::insert` calls `Expr::operator=` internally.
> 
> Ah, this could very likely be the reason (https://stackoverflow.com/a/3272731/12003165 says that "as of C++11, standard containers all use copy/move constructors for non-assignment operations (like push_back)", but doesn't say how vector's `operator=` gets implemented). Pre c++11 does require the value type to be copy-assignable/constructible (https://en.cppreference.com/w/cpp/container/vector)... It shouldn't be too difficult to construct a simple test case that tells whether `T::operator=` is used. If so, could you try another way to implement `exprs = b.exprs`?

Thanks very much for your comment! I'm willing to dig into this issue and try my best to solve it, but it may take some time as I find some build and test problems( Don't worry, I'll figure it out.

BTW, can this annoying issue be used as another argument that we should switch to Rust? (Hahaha, just kidding :rofl:

**Comment by k-ye** [comment]
I'd love to be able to use Rust :-) It's just that the tech stack (mainly the GPU stuff) Taichi depends on is not that mature

**Comment by junaire** [comment]
> I'd love to be able to use Rust :-) It's just that the tech stack (mainly the GPU stuff) Taichi depends on is not that mature 

Don't be too serious about this joke 🤣  I know nothing about rust, of course C++ is the same

**Comment by junaire** [comment]
Emmm... I try to mark `operator=` as `=delete` to see where it was called, and got tons of errors. It was like:

```
...

/home/jun/dev/cpp-projects/taichi/taichi/math/svd.h:115:16: error: overload resolution selected deleted operator '='
  Stiny_number = Tf(1.e-20f);

/home/jun/dev/cpp-projects/taichi/taichi/math/svd.h:118:8: error: overload resolution selected deleted operator '='
  Sa21 = load_if_ptr(a10);
...
```

That's little scary... I think it may be too hard to completely disable `Expr::operator=` for now. In my opinion, maybe we can offer two methods `set_or_assign` and `operator=`, like what we did in 5392e9e,  so the existing code wouldn't broke. 

What's your idea? @k-ye

**Comment by yuanming-hu** [comment]
Sorry about the mess in `svd.h` - that file was translated from a piece of SEE/AVX vectorized code to the legacy Taichi C++ frontend. I wrote a Python script for the translation two years ago. We may need to use regex to conduct batch replacement in that file. The replacement may need to go into a separate PR.

In the long run, since `svd.h` contains too many legacy C++ frontend constructs, maybe we should consider using the CHI IR builder to replace the machine-generated code, after we have Taichi function support. Note that the SVD function there emits a frontend AST and needs `lower_ast(...)`. WDYT? @k-ye (And maybe we need a standard library for functions like this...)

**Review by k-ye — DISMISSED** [review]
My bad. Didn't realize that this was a lot harder than expected. Thanks for pinning the problems!

> In my opinion, maybe we can offer two methods set_or_assign and operator=, like what we did in 5392e9e, so the existing code wouldn't broke.

Sounds great!

File: taichi/ir/expression.h [hunk 3]
```diff
@@ -71,22 +71,36 @@ class ExprGroup {
...
-    exprs.push_back(a);
-    exprs.push_back(b);
+    exprs.emplace_back(a);
+    exprs.emplace_back(b);
   }
 
   ExprGroup(const ExprGroup &a, const Expr &b) {
-    exprs = a.exprs;
-    exprs.push_back(b);
+    for (auto i = 0; i < a.size() + 1; i++) {
```
Comment (k-ye) [inline taichi/ir/expression.h:83]: This changes the code's logic, the right way is:

```cpp
exprs.resize(a.size() + 1);

for (int i = 0; i < a.size(); ++i) {
  exprs[i].set_or_insert_assignment(a.exprs[i]);
}
exprs.back().set_or_insert_assignment(b);

```

File: taichi/ir/expression.h [hunk 4]
```diff
@@ -71,22 +71,36 @@ class ExprGroup {
...
   ExprGroup(const Expr &a, const ExprGroup &b) {
-    exprs = b.exprs;
-    exprs.insert(exprs.begin(), a);
+    for (auto i = 0; i < b.size() + 1; i++) {
+      exprs.emplace_back();
+    }
+
+    exprs[0].set_or_insert_assignment(a);
+
+    for (auto i = 0; i < exprs.size(); i++) {
```
Comment (k-ye) [inline taichi/ir/expression.h:101]: ditto

**Review by k-ye — APPROVED** [review]
Sorry about the delay, LGTM!

**Pull request merged**